### PR TITLE
feat: add CompactSizeUint logic

### PR DESCRIPTION
**Pull request by tdelabro**

### CODE REVIEW

File: src/types/CompacSizeUint.zig [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    var buffer = [_]u8{0} ** 8;
+    std.mem.copyForwards(u8, &buffer, input[start..]);
+    switch (native_endian) {
+        .little => {},
+        .big => {
+            std.mem.reverse([8]u8, buffer);
+        },
+    }
+    return CompactSizeUint{ .inner = @bitCast(buffer) };
+}
```
Comment (tcoratger) [inline src/types/CompacSizeUint.zig:102]: Why not wrapping all of that inside a struct like:

```zig
pub const Hrp = struct {

const Self = @This();

pub fn decode(input: []const u8) !Self

....



};
```

File: src/types/CompacSizeUint.zig [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        std.mem.copyForwards(u8, res[1..], small_endian_value[0..]);
+        return res;
+    }
+}
+
+pub const DecodeCompactSizeUintError = enum {
+    EmptyInput,
+    InputTooLong,
+    InvalidInputLengthForPrefix,
+};
```
Comment (tcoratger) [inline src/types/CompacSizeUint.zig:65]: Here you should use a proper error like enum, see https://ziglang.org/documentation/master/#toc-Errors for ref. Should be something like

```zig
pub const DecodeCompactSizeUintError = error {
    EmptyInput,
    InputTooLong,
    InvalidInputLengthForPrefix,
};
```

File: src/types/CompacSizeUint.zig [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    var buffer = [_]u8{0} ** 8;
+    std.mem.copyForwards(u8, &buffer, input[start..]);
+    switch (native_endian) {
+        .little => {},
+        .big => {
+            std.mem.reverse([8]u8, buffer);
+        },
+    }
+    return CompactSizeUint{ .inner = @bitCast(buffer) };
```
Comment (tcoratger) [inline src/types/CompacSizeUint.zig:101]:
```suggestion
    return .{ .inner = @bitCast(buffer) };
```

File: src/types/CompacSizeUint.zig [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    const values = [_]u64{ 0, 252, 0xffff, 0xffffffff, std.math.maxInt(u64) };
+
+    for (values) |num| {
+        const allocator = std.testing.allocator;
+
+        const compact = CompactSizeUint.new(num);
+        const encoding = try compact.encode(allocator);
+        defer allocator.free(encoding);
+        const decoded = try CompactSizeUint.decode(encoding);
+        try std.testing.expect(decoded.value() == num);
```
Comment (tcoratger) [inline src/types/CompacSizeUint.zig:116]:
```suggestion
        try std.testing.expectEqual(num, decoded.value());
```

File: src/types/CompacSizeUint.zig [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+    for (0..1000) |_| {
+        const allocator = std.testing.allocator;
+        const num = rand.int(u64);
+
+        const compact = CompactSizeUint.new(num);
+        const encoding = try compact.encode(allocator);
+        defer allocator.free(encoding);
+        const decoded = try CompactSizeUint.decode(encoding);
+        try std.testing.expect(decoded.value() == num);
```
Comment (tcoratger) [inline src/types/CompacSizeUint.zig:131]:
```suggestion
        try std.testing.expectEqual(num, decoded.value());
```

File: src/types/CompacSizeUint.zig [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+const std = @import("std");
+const native_endian = @import("builtin").target.cpu.arch.endian();
+
+/// The inner value
+inner: u64,
+
+/// Returns a new instance
+pub inline fn new(inner: u64) CompactSizeUint {
+    return CompactSizeUint{ .inner = inner };
```
Comment (tcoratger) [inline src/types/CompacSizeUint.zig:20]:
```suggestion
    return .{ .inner = inner };
```

File: src/types/CompacSizeUint.zig (hunk 1 above)
Comment (tdelabro) [reply src/types/CompacSizeUint.zig:102]: The file is the struct. 
https://zig.news/gowind/zig-files-are-structs-288j

Don't hate the player, hate the game...

File: src/types/CompacSizeUint.zig (hunk 3 above)
Comment (tdelabro) [reply src/types/CompacSizeUint.zig:101]: done

File: src/types/CompacSizeUint.zig (hunk 4 above)
Comment (tdelabro) [reply src/types/CompacSizeUint.zig:116]: done

File: src/types/CompacSizeUint.zig (hunk 5 above)
Comment (tdelabro) [reply src/types/CompacSizeUint.zig:131]: done

File: src/types/CompacSizeUint.zig (hunk 6 above)
Comment (tdelabro) [reply src/types/CompacSizeUint.zig:20]: done

**Pull request merged**